### DISCUSSION TITLE
What's the role of PRNGKey in Haiku transforms?

### DISCUSSION THREAD

**Discussion by EngineerKhan** [jax-ml/jax#8252] [Q&A] [answered]
I have seen Haiku transforms where we have the provision of adding an optional PRNGKey. I wonder what job does it serve?

**Comment by tomhennigan** [comment, marked as answer]
In JAX prng keys are used to create random samples. You can find out more about them in [this guide](https://colab.research.google.com/github/google/jax/blob/main/docs/jax-101/05-random-numbers.ipynb).

In Haiku, `hk.transform` (and the variants of it) produce two functions: `init` and `apply`. Both of these functions accept random keys, and Haiku then makes new keys available inside your function via `hk.next_rng_key()`.

You can haiku given rng keys for any purpose you prefer, but typically rng keys are required for `init` since initial parameter values are random samples, and occasionally needed in `apply` when you need random samples as part of your forward pass (e.g. for [dropout](http://www.cs.toronto.edu/~hinton/absps/dropout.pdf) layers).

PS feel free to open haiku specific questions on the [dm-haiku github page](https://github.com/deepmind/dm-haiku).

**Reply by EngineerKhan** [reply]
>typically rng keys are required for init since initial parameter values are random samples

I have one follow-up question: Which distribution is used by default, and can we override it? I can't seem to see any such a thing, sorry.

**Reply by tomhennigan** [reply]
By default Haiku uses a truncated normal distribution with `stddev=1/sqrt(prod(fan_in))` for linear/conv. This should be documented for modules that ship with Haiku (e.g. see the [doc for `w_init` on hk.Linear](https://dm-haiku.readthedocs.io/en/latest/api.html#haiku.Linear)).

Most modules take a `w_init` (and `b_init` etc) argument in their constructor allowing you to override with your own initializer for each parameter. You can see an [example of us setting a different initializer in our transformer example](https://github.com/deepmind/dm-haiku/blob/3ce077d143dd07dbcbdeec884b5045d3f58a6113/examples/transformer/model.py#L61-L64).

Haiku ships with a number of [common initialisers](https://dm-haiku.readthedocs.io/en/latest/api.html?#module-haiku.initializers) that you can use (or you can write your own).

**Reply by EngineerKhan** [reply]
Thank you so much. One last question: does Flax (Linen) also work the same way? I have seen its `init()` also using the PRNG keys. I am especially curious about the Module class's `init()`.

**Reply by tomhennigan** [reply]
I am not very familiar with Flax (so I might be wrong here), but I suspect the RNG argument to `init` is for the same purpose.

FYI It is quite possible that they have chosen different default initialisers vs. Haiku (e.g. they may not use truncated normal for `nn.Dense`).

Flax modules seem to also support `{name}_init` args (e.g. `nn.Dense(.., kernel_init=X)`) and you can see one of [their transformer examples](https://github.com/google/flax/blob/19c4e273efefa644b694fd6984ee1c03a0a2a972/examples/lm1b/models.py#L189) overriding the default initializer.

**Reply by EngineerKhan** [reply]
I am really grateful for your detailed answers.
